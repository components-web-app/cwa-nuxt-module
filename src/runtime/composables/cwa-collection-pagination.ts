import { computed } from 'vue'

export type CwaPaginationProps = {
  currentPage: number
  totalPages: number
  maxPagesToDisplay: number
}

export type CwaPaginationEmits = {
  next: []
  previous: []
  change: [value: number]
}
export const useCwaCollectionPagination = (props: CwaPaginationProps) => {
  const pages = computed(() => {
    if (!props.totalPages) {
      return []
    }
    const allPages = Array.from(Array(props.totalPages), (_, x) => x + 1)
    const maxPagesToDisplay = props.maxPagesToDisplay || 7
    if (allPages.length < maxPagesToDisplay) {
      return allPages
    }

    const displayPages = []
    displayPages.push(props.currentPage)
    let lowest = props.currentPage
    let highest = props.currentPage
    let displayCounter = 1
    while (displayCounter < maxPagesToDisplay) {
      displayCounter++
      if ((displayCounter % 2 === 0 || highest >= props.totalPages) && lowest > 1) {
        lowest--
        displayPages.unshift(lowest)
        continue
      }
      if (highest < props.totalPages) {
        highest++
        displayPages.push(highest)
      }
    }
    return displayPages
  })

  return {
    pages,
  }
}
