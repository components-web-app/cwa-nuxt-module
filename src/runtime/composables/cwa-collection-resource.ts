import type { Ref } from 'vue'
import { computed, ref, watch } from 'vue'
import { type LocationQuery, useRoute } from 'vue-router'
import type { CwaResourceUtilsOps } from './cwa-resource'
import { useCwa, useCwaResource, useQueryBoundModel } from '#imports'
import type { CwaResource } from '#cwa/runtime/resources/resource-utils'

export const useCwaCollectionResource = (iri: Ref<string>, ops?: CwaResourceUtilsOps) => {
  const cwaResource = useCwaResource(iri, ops)
  const resource = cwaResource.getResource()

  const isLoadingCollection = ref(false)
  const fetchedCollectionItems = ref<CwaResource[]>()
  const loadCounter = ref(0)

  const $cwa = useCwa()
  const route = useRoute()
  // const { model: perPageModel } = useQueryBoundModel('perPage', { defaultValue: null, asNumber: true })
  const { model: pageModel } = useQueryBoundModel('page', { defaultValue: 1, asNumber: true })

  const collectionItems = computed<CwaResource[] | undefined>(() => {
    return fetchedCollectionItems.value || resource.value?.data?.collection?.['hydra:member']
  })

  const dataResourceIri = computed(() => {
    return resource.value?.data?.resourceIri
  })

  const totalPages = ref(1)

  function populateCollectionData(resource?: { collection?: { 'hydra:member': CwaResource[], 'hydra:view': { 'hydra:last': string } } } & CwaResource) {
    if (resource?.collection?.['hydra:member']) {
      fetchedCollectionItems.value = resource?.collection?.['hydra:member']
      const lastPagePath = resource?.collection?.['hydra:view']?.['hydra:last']
      if (!lastPagePath) {
        totalPages.value = 1
      }
      else {
        const urlParams = new URLSearchParams(lastPagePath.split('?')[1])
        const pageQueryParam = urlParams.get('page')
        totalPages.value = pageQueryParam ? (parseInt(pageQueryParam) || 1) : 1
      }
    }
  }

  async function reloadCollection() {
    if (!dataResourceIri.value) {
      return
    }
    const currentLoadCounter = ++loadCounter.value
    isLoadingCollection.value = true

    const { response } = $cwa.fetch({ path: iri.value })
    const { _data: resource } = await response

    if (currentLoadCounter === loadCounter.value) {
      populateCollectionData(resource)
      isLoadingCollection.value = false
    }
  }

  function goToNextPage() {
    if (pageModel.value >= totalPages.value) {
      return
    }
    changePage(pageModel.value + 1)
  }

  function goToPreviousPage() {
    if (!pageModel.value || pageModel.value <= 1) {
      return
    }
    changePage(pageModel.value - 1)
  }

  function changePage(newPageNumber: number) {
    pageModel.value = newPageNumber
  }

  // so components can be loaded in background still for the component manager to get metadata
  if (route) {
    watch(() => route.query, async (newQuery, oldQuery) => {
      const cleanPaginationFromQuery = (q: LocationQuery) => {
        const cleanQuery = { ...q }
        delete cleanQuery.perPage
        delete cleanQuery.page
        return cleanQuery
      }
      const cleanedOld = cleanPaginationFromQuery(oldQuery)
      const cleanedNew = cleanPaginationFromQuery(newQuery)
      if (JSON.stringify(cleanedOld) !== JSON.stringify(cleanedNew)) {
        pageModel.value = 1
      }

      if (JSON.stringify(oldQuery) === JSON.stringify(newQuery)) {
        return
      }
      await reloadCollection()
    })
  }

  populateCollectionData(resource.value?.data)

  watch(() => resource.value?.data, (newData) => {
    if (newData) {
      populateCollectionData(newData)
    }
  })

  return {
    ...cwaResource,
    resource,
    collectionItems,
    isLoadingCollection,
    pageModel,
    totalPages,
    goToNextPage,
    goToPreviousPage,
    changePage,
  }
}
