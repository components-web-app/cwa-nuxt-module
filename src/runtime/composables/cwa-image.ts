import { computed, onMounted, ref } from 'vue'
import type { Ref } from 'vue'
import { useCwaResource } from './cwa-resource'
import type { CwaResourceUtilsOps } from './cwa-resource'
import { useCwaResourceEndpoint } from './cwa-resource-endpoint'

export const useCwaImage = (iri: Ref<string>, imagineFilterName?: string, ops?: CwaResourceUtilsOps) => {
  type MediaFile = {
    contentUrl: string
    fileSize: number
    mimeType: string
    formattedFileSize: string
    imagineFilter?: string
    width?: number
    height?: number
  }

  const cwaResource = useCwaResource(iri, ops)
  const { query } = useCwaResourceEndpoint(iri)
  const resource = cwaResource.getResource()

  const loaded = ref(false)
  const image = ref()

  function handleLoad() {
    loaded.value = true
  }

  const imageFileMediaObjects = computed<MediaFile[] | undefined>(() => {
    return resource.value?.data?._metadata.mediaObjects?.file
  })

  const displayMedia = computed(() => {
    if (!imageFileMediaObjects.value || !imageFileMediaObjects.value.length) {
      return
    }
    const thumbnail = imageFileMediaObjects.value.filter(({ imagineFilter }) => (imagineFilter === imagineFilterName))
    return thumbnail?.[0] || imageFileMediaObjects.value[0]
  })

  const contentUrl = computed(() => {
    const mediaUrl = displayMedia.value?.contentUrl
    if (!mediaUrl) {
      return
    }
    return `${mediaUrl}${query.value}`
  })

  onMounted(() => {
    if (image.value?.complete || image.value?.naturalHeight !== 0) {
      handleLoad()
    }
  })

  return {
    ...cwaResource,
    contentUrl,
    displayMedia,
    handleLoad,
    loaded,
    resource,
  }
}
