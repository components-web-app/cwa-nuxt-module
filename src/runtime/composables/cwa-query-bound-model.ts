import { useRouter, useRoute, type LocationQueryValue } from 'vue-router'
import { computed, ref, watch } from 'vue'
import debounce from 'lodash-es/debounce'

type ModelOps = {
  defaultValue?: any
  delay?: number
  asNumber?: boolean
}

export const useQueryBoundModel = (queryParam: string | string[], ops?: ModelOps) => {
  const route = useRoute()
  const router = useRouter()
  let debounced: any

  const exp = new RegExp(`^${queryParam}\\[([a-zA-Z0-9]+)]$`, 'i')
  const matchingQueryParams = computed(() => {
    if (!route) {
      return []
    }
    if (Array.isArray(queryParam)) {
      const foundParams: string[] = []
      for (const p of queryParam) {
        route.query[p] && foundParams.push(p)
      }
      return foundParams
    }
    if (route.query[queryParam]) {
      return [queryParam]
    }
    const allQueryParams = Object.keys(route.query)
    return allQueryParams.filter(param => exp.test(param))
  })

  const matchedQueryParamValue = computed(() => {
    const matchingParams = matchingQueryParams.value
    if (!matchingParams || !matchingParams.length || !route) {
      return null
    }
    const normalizeValueAsArray = (valueIsArray: boolean, value: LocationQueryValue | LocationQueryValue[]) => {
      if (Array.isArray(value)) {
        return value
      }
      if (valueIsArray) {
        return [value]
      }
      return (ops?.asNumber && value !== null) ? Number(value) : value
    }

    if (Array.isArray(queryParam)) {
      for (const p of queryParam) {
        const valueIsArray = p.endsWith('[]')
        if (route.query[p]) {
          return normalizeValueAsArray(valueIsArray, route.query[p])
        }
      }
    }
    else if (route.query[queryParam]) {
      const valueIsArray = queryParam.endsWith('[]')
      return normalizeValueAsArray(valueIsArray, route.query[queryParam])
    }

    const matches = matchingParams[0].match(exp)
    if (!matches || matches.length < 2) {
      return null
    }
    const matchingObjectKey = matches[1]
    if (matchingParams.length && matchingObjectKey) {
      return { [matchingObjectKey]: route.query[matchingParams[0]] }
    }
    return null
  })

  const model = ref(matchedQueryParamValue.value !== null ? matchedQueryParamValue.value : ops?.defaultValue)

  watch(matchedQueryParamValue, (newValue) => {
    // if the query param value(s) changes, update the model
    model.value = newValue
  })

  watch(model, async (newValue) => {
    // update the query params on model change
    if (debounced) {
      debounced.cancel()
    }

    if (!route) {
      return
    }

    const filteredKeys = Object.keys(route.query).filter(key => !matchingQueryParams.value.includes(key))
    const newQuery: { [key: string]: LocationQueryValue | LocationQueryValue[] } = {}
    for (const retainedKey of filteredKeys) {
      newQuery[retainedKey] = route.query[retainedKey]
    }

    const isObj = Object.prototype.toString.call(newValue) === '[object Object]'
    if (isObj) {
      for (const newValueKey of Object.keys(newValue)) {
        if (!newValue[newValueKey]) {
          continue
        }
        if (Array.isArray(queryParam)) {
          for (const p of queryParam) {
            newQuery[`${p}[${newValueKey}]`] = newValue[newValueKey]
          }
        }
        else {
          newQuery[`${queryParam}[${newValueKey}]`] = newValue[newValueKey]
        }
      }
    }
    else if (newValue && (!Array.isArray(newValue) || newValue.length)) {
      if (Array.isArray(queryParam)) {
        for (const p of queryParam) {
          newQuery[p] = newValue
        }
      }
      else {
        newQuery[queryParam] = newValue
      }
    }

    debounced = debounce(async () => {
      await router.replace({ query: newQuery })
    }, ops?.delay || 10)
    await debounced()
  })

  return {
    model,
  }
}
