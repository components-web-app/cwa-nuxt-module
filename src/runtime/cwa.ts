import type { RouteLocationNormalizedLoaded, Router } from 'vue-router'
import type { CwaModuleOptions, CwaResourcesMeta } from '../module'
import { Storage } from './storage/storage'
import type { FetchEvent, FetchResourceEvent } from './api/fetcher/fetcher'
import Fetcher from './api/fetcher/fetcher'
import Mercure from './api/mercure'
import ApiDocumentation, { type ApiDocumentationComponentMetadataCollection } from './api/api-documentation'
import type { CwaApiDocumentationDataInterface } from './storage/stores/api-documentation/state'
import CwaFetch from './api/fetcher/cwa-fetch'
import FetchStatusManager from './api/fetcher/fetch-status-manager'
import { ResourcesManager } from './resources/resources-manager'
import { Resources } from './resources/resources'
import Auth from './api/auth'
import Forms from './api/forms'
import { useProcess } from './composables/process'
import Admin from './admin/admin'
import NavigationGuard from './admin/navigation-guard'
import { useRuntimeConfig, type NuxtApp } from '#app/nuxt'
import { useCookie } from '#app/composables/cookie.js'
import SiteConfig from '#cwa/runtime/api/site-config'

export default class Cwa {
  private readonly apiUrl: string
  private readonly options: CwaModuleOptions
  private readonly storage: Storage
  private readonly apiDocumentation: ApiDocumentation
  private readonly mercure: Mercure
  private readonly fetcher: Fetcher
  private readonly fetchStatusManager: FetchStatusManager
  private readonly cwaFetch: CwaFetch

  public readonly siteConfig: SiteConfig

  // public resources repository and utility getters
  public readonly resources: Resources

  // public service for managing resources - CRUD functions etc.
  public readonly resourcesManager: ResourcesManager

  // public service for authentication
  public readonly auth: Auth

  public readonly forms: Forms

  public readonly admin: Admin
  private readonly adminNavGuard: NavigationGuard

  public readonly currentModulePackageInfo: { version: string, name: string }

  constructor(nuxtApp: Pick<NuxtApp, '_route' | '_middleware' | '$router' | 'cwaResources'>, options: CwaModuleOptions, currentModulePackageInfo: { version: string, name: string }) {
    this.currentModulePackageInfo = currentModulePackageInfo
    const { isClient } = useProcess()
    const { public: { cwa: { apiUrl, apiUrlBrowser } } } = useRuntimeConfig()
    const defaultApiUrl = 'https://api-url-not-set.com'
    if (isClient) {
      this.apiUrl = apiUrlBrowser || apiUrl || defaultApiUrl
    }
    else {
      this.apiUrl = apiUrl || apiUrlBrowser || defaultApiUrl
    }

    this.cwaFetch = new CwaFetch(this.apiUrl)
    this.options = options
    this.storage = new Storage(this.options.storeName)
    this.siteConfig = new SiteConfig(this.cwaFetch, this.storage.stores.siteConfig)
    this.apiDocumentation = new ApiDocumentation(this.cwaFetch, this.storage.stores.apiDocumentation)
    this.mercure = new Mercure(this.storage.stores.mercure, this.storage.stores.resources, this.storage.stores.fetcher)
    this.fetchStatusManager = new FetchStatusManager(this.storage.stores.fetcher, this.mercure, this.apiDocumentation, this.storage.stores.resources)
    this.fetcher = new Fetcher(this.cwaFetch, this.fetchStatusManager, nuxtApp._route, this.storage.stores.resources)
    this.resources = new Resources(this.storage.stores.resources, this.storage.stores.fetcher)
    this.admin = new Admin(this.storage.stores.admin, this.storage.stores.resources, this.resources)
    this.resourcesManager = new ResourcesManager(this.cwaFetch, this.storage.stores.resources, this.fetchStatusManager, this.storage.stores.error, this.fetcher, this.admin, this.resources)
    this.auth = new Auth(
      this.cwaFetch,
      this.mercure,
      this.fetcher,
      this.admin,
      this.apiDocumentation,
      this.storage.stores.auth,
      this.storage.stores.resources,
      this.storage.stores.fetcher,
      useCookie('cwa_auth', { sameSite: 'strict' }),
    )
    this.forms = new Forms(this.storage.stores.resources)
    this.mercure.setFetcher(this.fetcher)
    this.mercure.setRequestCount(this.resourcesManager.requestCount)
    this.adminNavGuard = new NavigationGuard(nuxtApp.$router as Router, this.storage.stores.admin)
  }

  public get adminNavigationGuardFn() {
    return this.adminNavGuard.adminNavigationGuardFn
  }

  public get navigationDisabled() {
    return this.adminNavGuard.navigationDisabled
  }

  // API Documentation service is private, exposing only function required by applications
  public async getApiDocumentation(refresh = false): Promise<CwaApiDocumentationDataInterface | undefined> {
    return await this.apiDocumentation.getApiDocumentation(refresh)
  }

  public async getComponentMetadata(refresh = false, includePosition = false): Promise<undefined | ApiDocumentationComponentMetadataCollection> {
    return await this.apiDocumentation.getComponentMetadata(refresh, includePosition)
  }

  // fetcher is private, exposing the only function required by applications
  public fetchResource(event: FetchResourceEvent) {
    return this.fetcher.fetchResource(event)
  }

  public fetchRoute(route: RouteLocationNormalizedLoaded) {
    return this.fetcher.fetchRoute(route)
  }

  public fetch(event: FetchEvent) {
    return this.fetcher.fetch(event)
  }

  public clearPrimaryFetch() {
    this.fetchStatusManager.clearPrimaryFetch()
  }

  // Added as utility to bridge primary functionality of initialising 2 CWA services - this is not required by an application though, perhaps could be moved
  public async initClientSide() {
    await this.auth.init()
    this.mercure.init()
  }

  public get resourcesConfig(): CwaResourcesMeta {
    return this.options.resources || {}
  }

  // @internal
  public setResourceMeta(meta: CwaResourcesMeta) {
    this.options.resources = meta
  }

  public get layoutsConfig() {
    return this.options.layouts
  }

  public get pagesConfig() {
    return this.options.pages
  }

  public get pageDataConfig() {
    return this.options.pageData
  }

  public get appName(): string {
    return this.options.appName
  }

  public get apiUrlBase(): string {
    return this.apiUrl
  }
}
