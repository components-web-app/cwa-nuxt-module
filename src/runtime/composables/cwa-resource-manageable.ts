import {
  getCurrentInstance,
  onBeforeUnmount,
  onMounted,
  ref,

} from 'vue'
import type { ComponentPublicInstance, Ref } from 'vue'
import type { ManageableResourceOps } from '../admin/manageable-resource'
import ManageableResource from '../admin/manageable-resource'
import { useCwa } from './cwa'

/**
 * @internal
 * @description Advanced usage - usually this composable will be initialised from useCwaResource where disableManager does not equal true. Primarily separated for the ComponentGroup component
 */
export const useCwaResourceManageable = (iri: Ref<string | undefined>, ops?: ManageableResourceOps, proxy?: ComponentPublicInstance) => {
  const useProxy = proxy || getCurrentInstance()?.proxy
  if (!useProxy) {
    throw new Error(`Cannot initialise manager for resource. Instance is not defined with iri '${iri.value}'`)
  }
  const $cwa = useCwa()

  const manageableResource = new ManageableResource(useProxy, $cwa, ops || ref({}))

  onMounted(() => {
    manageableResource.init(iri)

    iri.value && $cwa.admin.eventBus.emit('componentMounted', iri.value)

    $cwa.admin.eventBus.on('manageableComponentMounted', (iriMounted) => {
      if (iriMounted === iri.value) {
        manageableResource.initNewIri()
        $cwa.admin.eventBus.emit('componentMounted', iri.value)
      }
    })
  })

  onBeforeUnmount(() => {
    manageableResource.clear()
  })

  return {
    manager: manageableResource,
  }
}
