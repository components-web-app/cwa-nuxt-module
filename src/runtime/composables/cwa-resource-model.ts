import { computed, getCurrentInstance, onBeforeUnmount, ref, watch } from 'vue'
import type { Ref } from 'vue'
import debounce from 'lodash-es/debounce'
import get from 'lodash-es/get'
import isObject from 'lodash-es/isObject'
import set from 'lodash-es/set'
import { useCwa } from '#cwa/runtime/composables/cwa'
import { useCwaResourceEndpoint } from '#cwa/runtime/composables/cwa-resource-endpoint'

interface ResourceModelOps {
  longWaitThreshold?: number
  debounceTime?: number
}

export const useCwaResourceModel = <T>(iri: Ref<string | undefined>, property: string | string[], ops?: ResourceModelOps) => {
  const proxy = getCurrentInstance()?.proxy
  const source = `input_${proxy?.$?.uid}`
  const $cwa = useCwa()
  const resource = computed(() => iri.value ? $cwa.resources.getResource(iri.value).value : undefined)

  const { endpoint } = useCwaResourceEndpoint(iri)

  const storeValue = computed<T | undefined>(() => (resource.value?.data ? get(resource.value.data, property) : undefined))
  const rootProperty = computed(() => {
    if (Array.isArray(property)) {
      return property[0]
    }
    return property.split('.')[0].split('[')[0]
  })
  const rootStoreValue = computed(() => (resource.value?.data ? get(resource.value.data, rootProperty.value) : undefined))

  const localValueWithIri = ref<{ [iri: string]: T | null | undefined }>({})
  const localValue = computed({
    get() {
      if (!iri.value) {
        return undefined
      }
      return localValueWithIri.value[iri.value]
    },
    set(newValue: T | null | undefined) {
      if (!iri.value) {
        return
      }
      localValueWithIri.value[iri.value] = newValue
    },
  })

  const pendingSubmit = ref(false)
  const isLongWait = ref(false)

  const longWaitThreshold = ops?.longWaitThreshold || 5000
  const debounceTime = ops?.debounceTime !== undefined ? ops.debounceTime : 250
  let debounced: any

  const isBusy = computed(() => pendingSubmit.value || isSubmitting.value)
  const isSubmitting = ref(false)
  const lastSubmittedValueInProgress = ref()

  function isEqual(value1: any, value2: any) {
    function requiresNormalizing(value: any) {
      const type = typeof value
      return value !== undefined && value !== null && (type === 'object' || Array.isArray(value))
    }

    function getNormalizedValue(value: any): string | null | undefined {
      return requiresNormalizing(value) ? JSON.stringify(value) : value
    }

    return getNormalizedValue(value1) === getNormalizedValue(value2)
  }

  async function updateResource(newLocalValue: any) {
    const submittingIri = iri.value
    if (!submittingIri) {
      return
    }

    // consider: another field is updating the same object and the request is not complete, we could overwrite that change with the new change as we are merging objects from the root property
    // consider: it doesn't matter if THIS field is updated and there is a pending request, we can continue and override that change
    // consider: it doesn't matter if another request is in progress for a DIFFERENT property, it'll update the store value and this request will not interfere
    // consider: it is all about the external sync status of a nested object property that we should need to wait for
    const isNewValueObject = isObject(newLocalValue)

    // todo: resolve the correct iri for the endpoint we are checking with the applied querystring -
    //  should be the same unless we will be creating a new draft in a request perhaps??
    //  ...but then would that matter, because the endpoint would be the same until then... to think about...
    isNewValueObject && await $cwa.resourcesManager.getWaitForRequestPromise(endpoint.value, rootProperty.value, source)

    // if we are already submitting this value, just skip - we shouldn't really get here
    // if the resource has been deleted, this will trigger a store value update but we do not need to submit this
    if (
      resource.value === undefined
      || (isSubmitting.value && isEqual(lastSubmittedValueInProgress.value, newLocalValue))
    ) {
      resetValue()
      return
    }

    // if we are submitting already, then we will already be updating the store value in a moment,
    // so the equality check would be invalid
    if (!isSubmitting.value && isEqual(storeValue.value, newLocalValue)) {
      pendingSubmit.value = false
      resetValue()
      return
    }

    isSubmitting.value = true
    let submittingValue
    // if updating a nested property within an object, we need to submit the object from the root, merging in the new value
    if (isNewValueObject) {
      const newObject = set({ [rootProperty.value]: { ...rootStoreValue.value } }, property, newLocalValue)
      submittingValue = newObject[rootProperty.value]
    }
    else {
      submittingValue = newLocalValue
    }
    lastSubmittedValueInProgress.value = submittingValue

    pendingSubmit.value = false
    const newResource = await $cwa.resourcesManager.updateResource({
      endpoint: endpoint.value,
      data: {
        [rootProperty.value]: submittingValue,
      },
      source,
    })
    isSubmitting.value = false

    // todo: check on when second request made before initial is complete
    const newIriReturned = newResource?.['@id'] && newResource['@id'] !== submittingIri
    if (newIriReturned) {
      localValueWithIri.value[newResource['@id']] = localValueWithIri.value[submittingIri]
      resetValue(submittingIri)
    }

    // we want to make sure the value is in sync with the store if we are not updating, which means resetting the locally stored variable
    // sequence of updating is:
    // local var > debounce > submitting var > store value
    isEqual(storeValue.value, submittingValue) && isEqual(storeValue.value, localValue.value) && resetValue()
  }

  function resetValue(resetIri?: string) {
    const useIri = resetIri || iri.value
    if (!useIri) {
      return
    }
    localValueWithIri.value[useIri] = undefined
  }

  const unwatchLocalValue = watch(localValue, (newLocalValue) => {
    if (newLocalValue === undefined) {
      return
    }
    if (debounced) {
      debounced.cancel()
    }
    pendingSubmit.value = true
    debounced = debounce(() => updateResource(newLocalValue), debounceTime)
    debounced()
  })

  let longWaitTimeoutFn: ReturnType<typeof setTimeout> | undefined

  const unwatchIsBusy = watch(isBusy, (newBusy) => {
    if (!newBusy) {
      isLongWait.value = false
      if (longWaitTimeoutFn) {
        clearTimeout(longWaitTimeoutFn)
      }
      return
    }

    longWaitTimeoutFn = setTimeout(() => {
      isLongWait.value = true
    }, longWaitThreshold)
  })

  const model = computed<T | undefined | null>({
    get() {
      if (localValue.value !== undefined) {
        return localValue.value
      }
      // when deleted, the store value is updating to undefined, then this model getter is null
      return storeValue.value || null
    },
    set(value) {
      localValue.value = value
    },
  })

  onBeforeUnmount(() => {
    unwatchLocalValue()
    unwatchIsBusy()
  })

  return {
    states: {
      pendingSubmit,
      submitting: isSubmitting,
      isBusy,
      isLongWait,
    },
    resetValue,
    model,
    localValueWithIri,
  }
}
