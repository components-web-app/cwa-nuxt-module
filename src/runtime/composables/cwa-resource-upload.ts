import { computed, ref } from 'vue'
import type { ComputedRef } from 'vue'
import { createConfirmDialog } from 'vuejs-confirm-dialog'
import { useCwaResourceEndpoint } from '#cwa/runtime/composables/cwa-resource-endpoint'
import { useCwa } from '#cwa/runtime/composables/cwa'
import ConfirmDialog from '#cwa/runtime/templates/components/core/ConfirmDialog.vue'

export const useCwaResourceUpload = (iri: ComputedRef<string | undefined>) => {
  const $cwa = useCwa()
  const resource = computed(() => iri.value ? $cwa.resources.getResource(iri.value).value : undefined)

  function getFilename() {
    return fileData.value ? `Existing Image (${fileData.value.formattedFileSize})` : ''
  }

  const fileData = computed(() => resource.value?.data?._metadata.mediaObjects?.file[0])

  const filenameInputModel = ref(getFilename())
  const fileExists = ref(true)
  const updating = ref(false)

  const { endpoint: updateEndpoint } = useCwaResourceEndpoint(iri, '/upload')
  const { endpoint: deleteEndpoint } = useCwaResourceEndpoint(iri)

  async function handleInputChangeFile(newFile: File | undefined) {
    if (!newFile || !iri.value) {
      return
    }
    updating.value = true
    const formData = new FormData()
    formData.append('file', newFile)
    await $cwa.resourcesManager.updateResource({
      iri: iri.value,
      endpoint: updateEndpoint.value,
      data: formData,
      headers: {
        accept: '*/*',
      },
    })
    filenameInputModel.value = getFilename()
    updating.value = false
  }

  async function confirmDelete() {
    const alertData = {
      title: 'Delete this image?',
      content: '<p>Are you sure you want to permanently delete this image?</p>',
    }
    // @ts-expect-error-next-line
    const dialog = createConfirmDialog(ConfirmDialog)
    const { isCanceled } = await dialog.reveal(alertData)

    return !isCanceled
  }

  async function handleInputDeleteFile() {
    if (!fileExists.value || !iri.value) {
      return
    }
    if (!await confirmDelete()) {
      return
    }
    updating.value = true
    await $cwa.resourcesManager.updateResource({
      endpoint: deleteEndpoint.value,
      data: {
        file: null,
      },
    })
    fileExists.value = false
    filenameInputModel.value = ''
    updating.value = false
  }

  return {
    filenameInputModel,
    updating,
    fileExists,
    handleInputChangeFile,
    handleInputDeleteFile,
  }
}
