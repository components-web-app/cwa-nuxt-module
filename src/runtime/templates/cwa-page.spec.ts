// @vitest-environment nuxt
import { describe, expect, test, vi } from 'vitest'
import { mount } from '@vue/test-utils'
import CwaPage from './cwa-page.vue'
import * as cwaComposable from '#cwa/runtime/composables/cwa'

const defaultIri = '12345'

describe('CWA page', () => {
  function createWrapper(iri = defaultIri) {
    vi.spyOn(cwaComposable, 'useCwa').mockImplementation(() => ({
      resources: {
        pageIri: {
          value: iri,
        },
      },
      siteConfig: {
        config: {
          fallbackTitle: true,
        },
      },
    }))

    return mount(CwaPage, {
      shallow: true,
    })
  }

  test('should display ResourceLoader component IF pageIri is defined', () => {
    const wrapper = createWrapper()
    const child = wrapper.findComponent({ name: 'ResourceLoader' })
    const { iri, componentPrefix } = child.props()

    expect(child).toBeDefined()
    expect(iri).toEqual(defaultIri)
    expect(componentPrefix).toEqual('CwaPage')
  })

  describe('snapshots', () => {
    test('should display ResourceLoader component IF pageIri is defined', () => {
      const wrapper = createWrapper()

      expect(wrapper.element).toMatchSnapshot()
    })

    test('should NOT display ResourceLoader IF pageIri is NOT defined', () => {
      const wrapper = createWrapper('')

      expect(wrapper.element).toMatchSnapshot()
    })
  })
})
