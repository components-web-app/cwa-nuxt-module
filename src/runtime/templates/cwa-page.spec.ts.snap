// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CWA page > snapshots > should NOT display ResourceLoader IF pageIri is NOT defined 1`] = `
<div
  class="cwa:page cwa:h-full"
>
  <!--v-if-->
</div>
`;

exports[`CWA page > snapshots > should display ResourceLoader component IF pageIri is defined 1`] = `
<div
  class="cwa:page cwa:h-full"
>
  <resourceloader-stub
    componentprefix="CwaPage"
    iri="12345"
  />
</div>
`;
